File: tadataka/vo/semi_dense/semi_dense.py
from warnings import warn

import numpy as np

from tqdm import tqdm

from tadataka.coordinates import image_coordinates
from tadataka.interpolation import interpolation
from tadataka.gradient import grad_x, grad_y
from tadataka.matrix import to_homogeneous, inv_motion_matrix, get_translation
from tadataka.rigid_transform import inv_transform
from tadataka.projection import inv_pi, pi
from tadataka.utils import is_in_image_range
from tadataka.vector import normalize_length
from tadataka.numeric import safe_invert
from tadataka.vo.semi_dense.depth import (
    calc_ref_inv_depth, calc_key_depth,
    InvDepthSearchRange, depth_search_range
)
from tadataka.vo.semi_dense.epipolar import (
    key_coordinates, key_epipolar_direction,
    ref_coordinates, ref_search_range
)
from tadataka.vo.semi_dense.flag import ResultFlag as FLAG
from tadataka.vo.semi_dense.gradient import GradientImage
from tadataka.vo.semi_dense._gradient import calc_gradient_norm
from tadataka.vo.semi_dense.variance import (
    calc_alpha, calc_observation_variance,
    photometric_variance, geometric_variance
)
from tadataka.vo.semi_dense.hypothesis import Hypothesis
from tadataka.vo.semi_dense._intensities import search_intensities
from tadataka.warp import warp2d, warp3d


def all_points_in_image(us, image_shape):
    return is_in_image_range(us, image_shape).all()


class InvDepthEstimator(object):
    def __init__(self, camera_model_key, image_key,
                 inv_depth_search_range, sigma_i, sigma_l,
                 step_size_ref, min_gradient):
        assert(np.ndim(image_key) == 2)
        assert(step_size_ref > 0)
        self.sigma_i = sigma_i
        self.sigma_l = sigma_l
        self.min_gradient = min_gradient
        self.camera_model_key = camera_model_key
        self.image_key = image_key
        self.step_size_ref = step_size_ref
        self.image_grad = GradientImage(grad_x(image_key), grad_y(image_key))
        self.inv_depth_range = InvDepthSearchRange(*inv_depth_search_range)

    def __call__(self, camera_model_ref, image_ref, T_rk, u_key, prior):
        if prior.inv_depth <= 0:
            return prior, FLAG.NEGATIVE_PRIOR_DEPTH

        inv_depth_range = self.inv_depth_range(prior)
        if inv_depth_range is None:
            return prior, FLAG.HYPOTHESIS_OUT_OF_SERCH_RANGE

        x_key = self.camera_model_key.normalize(u_key)

        # step size / inv depth = approximately const
        inv_depth_ref = calc_ref_inv_depth(T_rk, x_key, prior.inv_depth)
        step_size_key = (prior.inv_depth / inv_depth_ref) * self.step_size_ref
        if inv_depth_ref <= 0:
            return prior, FLAG.NEGATIVE_REF_DEPTH

        xs_key = key_coordinates(get_translation(T_rk), x_key, step_size_key)
        us_key = self.camera_model_key.unnormalize(xs_key)
        if not all_points_in_image(us_key, self.image_key.shape):
            return prior, FLAG.KEY_OUT_OF_RANGE

        intensities_key = interpolation(self.image_key, us_key)
        gradient_key = calc_gradient_norm(intensities_key)

        if gradient_key < self.min_gradient:
            return prior, FLAG.INSUFFICIENT_GRADIENT

        x_range_ref = ref_search_range(T_rk, x_key,
                                       depth_search_range(*inv_depth_range))
        xs_ref = ref_coordinates(x_range_ref, self.step_size_ref)
        if len(xs_ref) < len(xs_key):
            return prior, FLAG.REF_EPIPOLAR_TOO_SHORT

        us_ref = camera_model_ref.unnormalize(xs_ref)

        if not is_in_image_range(us_ref[0], image_ref.shape):
            return prior, FLAG.REF_CLOSE_OUT_OF_RANGE

        # TODO when does this condition become true?
        if not is_in_image_range(us_ref[-1], image_ref.shape):
            return prior, FLAG.REF_FAR_OUT_OF_RANGE

        intensities_ref = interpolation(image_ref, us_ref)
        argmin = search_intensities(intensities_key, intensities_ref)
        depth_key = calc_key_depth(T_rk, x_key, xs_ref[argmin])

        x_max_ref, x_min_ref = x_range_ref
        direction = normalize_length(x_max_ref - x_min_ref)
        variance = calc_observation_variance(
            alpha=calc_alpha(T_rk, x_key, direction, prior.inv_depth),
            geo_variance=geometric_variance(
                key_epipolar_direction(get_translation(T_rk), x_key),
                self.image_grad(u_key), self.sigma_l, epsilon=1e-4),
            photo_variance=photometric_variance(gradient_key / step_size_key,
                                                self.sigma_i)
        )
        return Hypothesis(safe_invert(depth_key), variance), FLAG.SUCCESS


class AgeDependentValues(object):
    def __init__(self, age_map, values):
        self.age_map = age_map
        self.values = values

    def __call__(self, u):
        x, y = u
        age = self.age_map[y, x]
        if age == 0:
            return None

        return self.values[-age]


class InvDepthMapEstimator(object):
    def __init__(self, estimator):
        self._estimator = estimator

    def __call__(self, inv_depth_map, variance_map,
                 age_map, refframes):
        age_dict = AgeDependentValues(age_map, refframes)
        assert(inv_depth_map.shape == variance_map.shape == age_map.shape)

        flag_map = np.full(inv_depth_map.shape, FLAG.NOT_PROCESSED)
        for u_key in tqdm(image_coordinates(inv_depth_map.shape)):
            ref = age_dict(u_key)
            if ref is None:
                continue

            x, y = u_key
            prior = Hypothesis(inv_depth_map[y, x], variance_map[y, x])
            result, flag = self._estimator(*ref, u_key, prior)

            inv_depth_map[y, x] = result.inv_depth
            variance_map[y, x] = result.variance
            flag_map[y, x] = flag

        return inv_depth_map, variance_map, flag_map

File: src/semi_dense/semi_dense.rs
use std::convert::TryInto;

use indicatif::ProgressBar;
use ndarray::{arr1, Array, Array1, Array2};
use ndarray_linalg::solve::Inverse;
use ndarray_linalg::Norm;

use crate::camera::Normalizer;
use crate::gradient::gradient1d;
use crate::interpolation::Interpolation;
use crate::image_range::{ImageRange, all_in_range};
use crate::transform::get_translation;
use crate::warp::Warp;
use super::depth::{calc_key_depth, calc_ref_depth, depth_search_range};
use super::epipolar::{key_coordinates, ref_coordinates};
use super::flag::Flag;
use super::frame::Frame;
use super::gradient::ImageGradient;
use super::hypothesis;
use super::hypothesis::Hypothesis;
use super::intensities;
use super::numeric::Inv;
use super::numeric::Inverse as InverseDepth;
use super::params::Params;
use super::variance::{calc_alpha, calc_variance, geo_var, photo_var};

fn step_ratio(
    transform_rk: &Array2<f64>,
    x_key: &Array1<f64>,
    key_inv_depth: Inv,
) -> Result<f64, Flag> {
    let key_depth: f64 = key_inv_depth.inv();
    let ref_depth = calc_ref_depth(transform_rk, x_key, key_depth);
    if ref_depth <= 0. {
        return Err(Flag::NegativeRefDepth);
    }
    let ref_inv_depth = ref_depth.inv();
    let ratio = key_inv_depth / ref_inv_depth;
    Ok(f64::from(ratio))
}

fn xs_key(
    transform_rk: &Array2<f64>,
    x_key: &Array1<f64>,
    key_step_size: f64
) -> Array2<f64> {
    let t_rk = get_translation(transform_rk);
    key_coordinates(&t_rk, x_key, key_step_size)
}

fn xs_ref(
    transform_rk: &Array2<f64>,
    x_key: &Array1<f64>,
    depth_range: (f64, f64),
    ref_step_size: f64
) -> Array2<f64> {
    let (min_depth, max_depth) = depth_range;
    let (x_min_ref, _) = transform_rk.warp(x_key, min_depth);
    let (x_max_ref, _) = transform_rk.warp(x_key, max_depth);
    ref_coordinates(&x_min_ref, &x_max_ref, ref_step_size)
}

fn check_us_ref(
    us_ref: &Array2<f64>,
    us_key_size: usize,
    ref_image_shape: &[usize]
) -> Result<(), Flag> {
    if us_ref.nrows() < us_key_size {
        return Err(Flag::RefEpipolarTooShort);
    }

    if !us_ref.row(0).is_in_range(ref_image_shape) {
        return Err(Flag::RefCloseOutOfRange);
    }

    // TODO when does this condition become true?
    if !us_ref.row(us_ref.nrows()-1).is_in_range(ref_image_shape) {
        return Err(Flag::RefFarOutOfRange);
    }

    Ok(())
}

fn transform_rk(
    transform_wk: &Array2<f64>,
    transform_wr: &Array2<f64>,
) -> Array2<f64> {
    let transform_rw = transform_wr.inv().unwrap();
    transform_rw.dot(transform_wk)
}

pub fn estimate(
    u_key: &Array1<f64>,
    prior: &Hypothesis,
    keyframe: &Frame,
    refframe: &Frame,
    image_grad: &ImageGradient,
    params: &Params,
) -> Result<Hypothesis, Flag> {
    let transform_rk = transform_rk(&keyframe.transform, &refframe.transform);

    let depth_range = depth_search_range(&prior.range());
    let x_key = keyframe.camera_params.normalize(u_key);

    // calculate step size along the epipolar line on the keyframe
    // step size / inv depth = approximately const
    let result = step_ratio(&transform_rk, &x_key, prior.inv_depth);
    let key_step_size = match result {
        Err(e) => return Err(e),
        Ok(ratio) => ratio * params.ref_step_size,
    };

    // calculate coordinates on the keyframe image
    let xs_key = xs_key(&transform_rk, &x_key, key_step_size);
    let us_key = keyframe.camera_params.unnormalize(&xs_key);
    if !all_in_range(&us_key, keyframe.image.shape()) {
        return Err(Flag::KeyOutOfRange);
    }

    // extract intensities from the key coordinates
    let key_intensities = keyframe.image.interpolate(&us_key);
    let key_gradient = gradient1d(&key_intensities).norm();
    // most of coordinates has insufficient gradient
    // return early to reduce computation
    if key_gradient < params.min_gradient {
        return Err(Flag::InsufficientGradient);
    }

    // calculate coordinates on the reference frame image
    let xs_ref = xs_ref(&transform_rk, &x_key, depth_range, params.ref_step_size);
    let us_ref = refframe.camera_params.unnormalize(&xs_ref);
    check_us_ref(&us_ref, us_key.nrows(), refframe.image.shape())?;

    // extract intensities from the ref coordinates
    let ref_intensities = refframe.image.interpolate(&us_ref);

    // search along epipolar line and calculate depth
    let argmin = intensities::search(&ref_intensities, &key_intensities);
    let key_depth = calc_key_depth(&transform_rk, &x_key, &xs_ref.row(argmin));
    // calculate variance
    let alpha = calc_alpha(&transform_rk, &x_key, depth_range, key_depth);
    let t_rk = get_translation(&transform_rk);
    let geo_var = geo_var(&x_key, &t_rk, &image_grad.get(&u_key));
    let photo_var = photo_var(key_gradient / key_step_size);
    let variance = calc_variance(alpha, geo_var, photo_var, &params.var_coeffs);

    hypothesis::check_args(key_depth.inv(), variance, params.inv_depth_range)?;
    Ok(Hypothesis::new(key_depth.inv(), variance, params.inv_depth_range))
}

pub fn update_depth(
    keyframe: &Frame,
    refframes: &Vec<Frame>,
    age_map: &Array2<usize>,
    prior_depth: &Array2<f64>,
    prior_variance: &Array2<f64>,
    params: &Params,
) -> (Array2<i64>, Array2<f64>, Array2<f64>) {
    assert!(age_map.shape() == prior_depth.shape());
    assert!(age_map.shape() == prior_variance.shape());
    assert!(age_map.shape() == keyframe.image.shape());
    for i in 0..refframes.len() {
        assert!(age_map.shape() == refframes[i].image.shape());
    }

    let image_grad = ImageGradient::new(&keyframe.image);
    let height = keyframe.image.shape()[0];
    let width = keyframe.image.shape()[1];

    let mut flag_map = Array::zeros((height, width));
    let mut result_depth = Array::zeros((height, width));
    let mut result_variance = Array::zeros((height, width));
    let inv_depth_range = params.inv_depth_range;

    let bar = ProgressBar::new((height * width).try_into().unwrap());

    for y in 0..height {
        for x in 0..width {
            bar.inc(1);

            let age = age_map[[y, x]];
            let d = prior_depth[[y, x]];
            let v = prior_variance[[y, x]];

            if age == 0 {
                // refframe cannot be observed from this pixel
                result_depth[[y, x]] = d;
                result_variance[[y, x]] = v;
                flag_map[[y, x]] = Flag::NotProcessed as i64;
                continue;
            }

            if refframes.len() < age {
                eprintln!("Age exceeds the refframe size");
                std::process::exit(1);
            }

            let refframe = &refframes[refframes.len()-age];

            if let Err(f) = hypothesis::check_args(d.inv(), v, inv_depth_range) {
                result_depth[[y, x]] = d;
                result_variance[[y, x]] = v;
                flag_map[[y, x]] = f as i64;
                continue;
            }

            let u_key = arr1(&[x as f64, y as f64]);
            let prior = Hypothesis::new(d.inv(), v, inv_depth_range);
            let result = estimate(&u_key, &prior, &keyframe, refframe,
                                  &image_grad, &params);
            let (h, flag) = match result {
                Err(flag) => (prior, flag),
                Ok(h) => (h, Flag::Success),
            };

            result_depth[[y, x]] = h.inv_depth.inv();
            result_variance[[y, x]] = h.variance;
            flag_map[[y, x]] = flag as i64;
        }
    }

    bar.finish();

    (flag_map, result_depth, result_variance)
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::arr2;

    #[test]
    fn test_step_ratio() {
        let transform_rk = arr2(
            &[[0., 0., 1., -2.],
              [0., 1., 0., 2.],
              [-1., 0., 0., 7.],
              [0., 0., 0., 1.]]
        );
        let key_depth = 2.0;
        let x_key = arr1(&[2.0, 2.0]);
        // ref_depth = -(2.0 * 2.0) + 7.0 = 3.0
        let ref_depth = 3.0;

        approx::assert_abs_diff_eq!(
            step_ratio(&transform_rk, &x_key, key_depth.inv()).unwrap(),
            (1. / key_depth) / (1. / ref_depth),
            epsilon = 1e-10
        );

        let key_depth = 4.0;
        let x_key = arr1(&[2.0, 2.0]);
        // ref_depth = -(4.0 * 2.0) + 7.0 = -1.0
        assert_eq!(
            step_ratio(&transform_rk, &x_key, key_depth.inv()).unwrap_err(),
            Flag::NegativeRefDepth
        );
    }

    #[test]
    fn test_check_us_ref() {
        let ref_image_shape: [usize; 2] = [40, 30];
        let us_key_size = 2;

        let us_ref = arr2(
            &[[10., 20.],
              [0., 0.]]
        );
        let result = check_us_ref(&us_ref, us_key_size, &ref_image_shape);
        assert_eq!(result.unwrap(), ());

        let us_ref = arr2(&[[10., -2.]]);
        let result = check_us_ref(&us_ref, us_key_size, &ref_image_shape);
        assert_eq!(result.unwrap_err(), Flag::RefEpipolarTooShort);

        let us_ref = arr2(
            &[[10., -2.],
              [0., 0.]]
        );
        let result = check_us_ref(&us_ref, us_key_size, &ref_image_shape);
        assert_eq!(result.unwrap_err(), Flag::RefCloseOutOfRange);

        let us_ref = arr2(
            &[[0., 0.],
              [10., -2.]]
        );
        let result = check_us_ref(&us_ref, us_key_size, &ref_image_shape);
        assert_eq!(result.unwrap_err(), Flag::RefFarOutOfRange);
    }
}
